pkg/tempo/automaticloggingprocessor/automaticloggingprocessor.go

```diff
@@ -0,0 +1,248 @@
+				span := ils.Spans().At(k)
+				traceID := span.TraceID().HexString()
+
+				if p.cfg.Spans {
+					p.exportToLoki(typeSpan, traceID, p.spanKeyVals(span, svc)...)
+				}
+
+				if p.cfg.Roots && span.ParentSpanID().IsEmpty() {
+					p.exportToLoki(typeRoot, traceID, p.spanKeyVals(span, svc)...)
+				}
+
+				if p.cfg.Processes && lastTraceID != traceID {
+					lastTraceID = traceID
+					p.exportToLoki(typeProcess, traceID, p.processKeyVals(rs.Resource(), svc)...)
+				}
+			}
+		}
+	}
+
+	return p.nextConsumer.ConsumeTraces(ctx, td)
+}
+
+func (p *automaticLoggingProcessor) GetCapabilities() component.ProcessorCapabilities {
+	return component.ProcessorCapabilities{}
+}
+
+// Start is invoked during service startup.
+func (p *automaticLoggingProcessor) Start(ctx context.Context, _ component.Host) error {
+	loki := ctx.Value(contextkeys.Loki).(*loki.Loki)
+	if loki == nil {
+		return fmt.Errorf("key does not contain a Loki instance")
+	}
+	p.lokiInstance = loki.Instance(p.cfg.LokiName)
+	if p.lokiInstance == nil {
+		return fmt.Errorf("loki instance %s not found", p.cfg.LokiName)
+	}
+	return nil
+}
+
+// Shutdown is invoked during service shutdown.
+func (p *automaticLoggingProcessor) Shutdown(context.Context) error {
+	p.done.Store(true)
+
+	return nil
+}
+
+func (p *automaticLoggingProcessor) processKeyVals(resource pdata.Resource, svc string) []interface{} {
+	atts := make([]interface{}, 0, 2) // 2 for service name
+	rsAtts := resource.Attributes()
+
+	// name
+	atts = append(atts, p.cfg.Overrides.ServiceKey)
+	atts = append(atts, svc)
+
+	for _, name := range p.cfg.ProcessAttributes {
+		att, ok := rsAtts.Get(name)
+		if ok {
+			// name/key val pairs
+			atts = append(atts, name)
+			atts = append(atts, attributeValue(att))
+		}
+	}
+
+	return atts
+}
+
+func (p *automaticLoggingProcessor) spanKeyVals(span pdata.Span, svc string) []interface{} {
+	atts := make([]interface{}, 0, 8) // 8 for name, duration and service name
+
+	atts = append(atts, p.cfg.Overrides.SpanNameKey)
+	atts = append(atts, span.Name())
+
+	atts = append(atts, p.cfg.Overrides.DurationKey)
+	atts = append(atts, spanDuration(span))
+
+	atts = append(atts, p.cfg.Overrides.ServiceKey)
+	atts = append(atts, svc)
+
+	atts = append(atts, p.cfg.Overrides.StatusKey)
+	atts = append(atts, span.Status().Code())
+
+	span.Status().Code()
+
+	for _, name := range p.cfg.SpanAttributes {
+		att, ok := span.Attributes().Get(name)
+		if ok {
+			atts = append(atts, name)
+			atts = append(atts, attributeValue(att))
+		}
+	}
+
+	return atts
+}
+
+func (p *automaticLoggingProcessor) exportToLoki(kind string, traceID string, keyvals ...interface{}) {
+	if p.done.Load() {
+		return
+	}
+
+	keyvals = append(keyvals, []interface{}{p.cfg.Overrides.TraceIDKey, traceID}...)
+	line, err := logfmt.MarshalKeyvals(keyvals...)
+	if err != nil {
+		level.Warn(p.logger).Log("msg", "unable to marshal keyvals", "err", err)
+		return
+	}
+
+	sent := p.lokiInstance.SendEntry(api.Entry{
```

Comment:
In the future we'll have to be careful to make sure the Loki instances don't get replaced so this doesn't end up with a stale reference. I think we're fine for now though.

We could be defensive and perform the lookup each time, but I'm not sure it's worth it just yet.